feat: add a `detach` mode to the CLI (#3568)
trufflesuite/ganache-cli: src/packages/ganache/README.md

  -p, --server.port, --port             Port to listen on.
                                        deprecated aliases: --port
                                                                                      [number] [default: 8545]
```

### Detached Instances

Ganache can be started as a background instance via the CLI by providing the following argument (along with any combination
of the Ganache startup arguments above):

```
  -D, --detach, --😈                     Run Ganache in detached (daemon) mode.                       [boolean]
```

This will start Ganache as a background process, and return to the console as soon as Ganache has started and ready to
receive requests. A friendly name will be returned to STDOUT which can then be used to interact with the instance via
the `ganache instances` command with the following arguments:

```
Commands:
  cli.ts instances list         List instances running in detached mode

>   cli.ts instances stop <name>  Stop the instance specified by <name>

Reviewer: ```suggestion
  ganache instances list         List instances running in detached mode
  ganache instances stop <name>  Stop the instance specified by <name>
```

  -p, --server.port, --port             Port to listen on.
                                        deprecated aliases: --port
                                                                                      [number] [default: 8545]
```

### Detached Instances

Ganache can be started as a background instance via the CLI by providing the following argument (along with any combination
of the Ganache startup arguments above):

```
  -D, --detach, --😈                     Run Ganache in detached (daemon) mode.                       [boolean]
```

> This will start Ganache as a background process, and return to the console as soon as Ganache has started and ready to

Reviewer 2: ```suggestion
This will start Ganache as a background process, and return to the console as soon as Ganache has started and is ready to
```

receive requests. A friendly name will be returned to STDOUT which can then be used to interact with the instance via
the `ganache instances` command with the following arguments:

```

> Commands:

Reviewer 2: This is cool. 

Is it possible/worth an alias from `ganache instances start [params]` to `ganache [params] --detach` ?

Author: I think that's a great idea. I found myself typing `ganache instances...` when trying to start a new instance, so this would definitely be an improvement.

I made https://github.com/trufflesuite/ganache/issues/3903 to cover that.

  ganache instances list         List instances running in detached mode
  ganache instances stop <name>  Stop the instance specified by <name>
```

E.g., start Ganache on port 8544, with a block time of 10 seconds, and then stop the instance.

```
$ ganache --port=8544 --miner.blockTime=10 --detach
salted_caramel_ganache

$ ganache instances list
╔═══════╤════════════════════════╤══════════╤═════════╤═══════════╤══════╤════════════╗
║ PID   │ Name                   │ Flavor   │ Version │ Host      │ Port │ Uptime     ║
╟───────┼────────────────────────┼──────────┼─────────┼───────────┼──────┼────────────╢
║ 56004 │ salted_caramel_ganache │ ethereum │ 7.4.3   │ 127.0.0.1 │ 8544 │ 48 seconds ║
╚═══════╧════════════════════════╧══════════╧═════════╧═══════════╧══════╧════════════╝

$ ganache instances stop salted_caramel_ganache

> Process stopped

Reviewer 3: Can you add the `With the following command, you can start Ganache, run your tests, and stop Ganache when you are finished.` example to the README?

  -p, --server.port, --port             Port to listen on.
                                        deprecated aliases: --port
                                                                                      [number] [default: 8545]
```

### Detached Instances

> Ganache can be started as a background instance via the CLI by providing the following argument (along with any combination

Reviewer 3: ```suggestion
Ganache can be started as a background instance via the CLI by providing the following argument (along with any valid combination
```

of the Ganache startup arguments above):

> ```

Reviewer 3: ```suggestion
```console
```

Reviewer 3: Note: i'm not 100% sure when to use `console` and when to use `bash` for this stuff.

Author: So I had a look around, and take this with a grain of salt, because I don't have anything authoritative,  but:

https://github.com/github/linguist/blob/3c3b037910006fc2f1a9bb34b2c4e9cde062206c/lib/linguist/languages.yml shows the languages that github knows, and the various aliases.

- `bash` maps to `Shell`
- `console` maps to `ShellSession`

I believe that the difference is that a `Shell` block should be a shell _script_, where as a `ShellSession` should be a series of commands and responses from a session.

Given all of these examples are in the context of a _session_, I've used `console`. 

While I'm here, I standardised the console blocks, as some are missing the `$` prompt, which is somewhat confusing (at the very least, inconsistent!)

  -D, --detach, --😈                     Run Ganache in detached (daemon) mode.                       [boolean]
```

This will start Ganache as a background process, and return to the console as soon as Ganache has started and is ready to
receive requests. A friendly name will be returned to STDOUT which can then be used to interact with the instance via
the `ganache instances` command with the following arguments:

> ```

Reviewer 3: ```suggestion
```console
```

Commands:
  ganache instances list         List instances running in detached mode
  ganache instances stop <name>  Stop the instance specified by <name>
```

E.g., start Ganache on port 8544, with a block time of 10 seconds, and then stop the instance.

> ```

Reviewer 3: ```suggestion
```console
```

$ ganache --port=8544 --miner.blockTime=10 --detach
salted_caramel_ganache

$ ganache instances list
╔═══════╤════════════════════════╤══════════╤═════════╤═══════════╤══════╤════════════╗
║ PID   │ Name                   │ Flavor   │ Version │ Host      │ Port │ Uptime     ║
╟───────┼────────────────────────┼──────────┼─────────┼───────────┼──────┼────────────╢
║ 56004 │ salted_caramel_ganache │ ethereum │ 7.4.3   │ 127.0.0.1 │ 8544 │ 48 seconds ║
╚═══════╧════════════════════════╧══════════╧═════════╧═══════════╧══════╧════════════╝

$ ganache instances stop salted_caramel_ganache
Process stopped
```

With the following command, you can start Ganache, run your tests, and stop Ganache when you are finished.

> ```

Reviewer 3: ```suggestion
```bash
```

Author: 👍, except I used `console` as per my justification above.

GANACHE=$(ganache --detach) && npm run test; ganache instances stop $GANACHE
```

> Or if you are running Powershell on Windows, you can do:

Reviewer 3: ```suggestion
Or if you are running PowerShell on Windows, you can do:
```

```

> $GANACHE=ganache --detach; npm run test; ganache instances stop $GANACHE; Remove-Variable GANACHE

Reviewer 3: ```suggestion
```PowerShell
GANACHE=ganache --detach; npm run test; ganache instances stop $GANACHE; Remove-Variable GANACHE
```
